File: Java Implementation/Node.java
import java.util.HashMap;
import java.util.Set;

public class Node {

	private int id;
	private HashMap<Integer, Double> connections;
	private HashMap<Integer, Double> strengths;
	
	public Node(int id) {
		this.id = id;
		this.connections = new HashMap<Integer, Double>();
		this.strengths = new HashMap<Integer, Double>();
	}
	
	Node(Node n){
		id = n.id;
		connections = n.connections;
		strengths = n.strengths;
	}
	
	public int getId() {
		return this.id;
	}
	
	public String toString() {
		return Integer.toString(this.id);
	}
	
	public HashMap<Integer, Double> neighbors(){
		HashMap<Integer, Double> neighbors = new HashMap<Integer, Double>();
		Set<Integer> keys = connections.keySet();
		for(int key : keys) {
			neighbors.put(key, connections.get(key));
		}
		return neighbors;
	}
	
	public void addConnection(Node node, double value) {
		this.connections.put(node.getId(), value);
	}

	public void addConnection(Node node) {
		this.connections.put(node.getId(), 1.0);
	}
	
	public double edgeWeight(int nodeId) {
		if(this.connections.containsKey(nodeId)) {
			return this.connections.get(nodeId);
		}
		return 0.0;
	}
	
	public double edgeWeight(Node node) {
		return this.edgeWeight(node.getId());
	}
	
	public void removeNeighbor(Node node) {
		this.removeNeighbor(node.getId());
	}

	public void removeNeighbor(int nodeId) {
		this.connections.remove(nodeId);
	}
	
	public boolean hasEdge(Node node) {
		return this.hasEdge(node.getId());
	}
	
	public boolean hasEdge(int nodeId) {
		return this.connections.containsKey(nodeId);
}
	
	public void assignStrength(int neighborId, double strength) {
		this.strengths.put(neighborId, strength);
	}
	
	public double getStrength(int neighborId) {
		if(this.strengths.containsKey(neighborId)) {
			return this.strengths.get(neighborId);
		}
		return 0.0;
	}
	
	public void showConnections() {
		System.out.println(this.connections);
	}
	
	public void showStrengths() {
		System.out.println(this.strengths);
	}
}

File: C++ Implementation/Node.cpp
#include "Node.hpp"


Node::Node(int _id) : id(_id) {
    this->connections = new std::unordered_map<Node*, float>;
    this->strengths = new std::unordered_map<Node*, float>;
}

Node::Node(const Node &_node): id(_node.id), connections(_node.connections), strengths(_node.strengths) {}

Node::~Node() {
    delete connections;
    delete strengths;
}

int Node::get_id(){
    return this->id;
}

float Node::degree() {
    float degree = 0.0;
    auto it = connections->find(this);
    if(it != connections->end()){
        degree += it->second;
    }
    for(auto neighbor : *connections){
        degree += neighbor.second;
    }
    return degree;
}

std::unordered_map<Node*, float>* Node::neighbors(){
    return this->connections;
}

void Node::add_connection(Node *node, float value){
    if(this->connections->find(node) != this->connections->end()) {
        this->connections->at(node) = value;
        return;
    }
    this->connections->insert(std::make_pair(node, value));
}

void Node::add_connection(Node *node){
    if(this->connections->find(node) != this->connections->end()) {
        this->connections->at(node) = 1.0;
        return;
    }
    this->connections->insert(std::make_pair(node, 1.0));
}

void Node::assign_strength(Node* neighbor, float strength_value) {
    if(this->strengths->find(neighbor) != this->strengths->end()) {
        this->strengths->at(neighbor) = strength_value;
        return;
    }
    this->strengths->insert(std::make_pair(neighbor, strength_value));
}

float Node::get_strength(Node* neighbor) {
    auto it = this->strengths->find(neighbor);
    if(it != this->strengths->end()) {
        return it->second;
    }
    return 0.0;
}

bool Node::has_edge(Node* node) {
    return this->connections->find(node) != this->connections->end();
}

float Node::get_edge_weight(Node* node) const{
    auto it = this->connections->find(node);
    if(it != this->connections->end()){
        return it->second;
    }
    return 0.0;
}

void Node::remove_neighbor(Node* node){
    this->connections->erase(node);
}
